versioned/persist/dynamodb/src/main/java/org/projectnessie/versioned/persist/dynamodb/DynamoDatabaseClient.java

```diff
@@ -0,0 +1,156 @@
+import static org.projectnessie.versioned.persist.dynamodb.Tables.TABLE_KEY_LISTS;
+
+import java.net.URI;
+import java.util.List;
+import java.util.stream.Stream;
+import org.projectnessie.versioned.persist.adapter.DatabaseConnectionProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.http.urlconnection.UrlConnectionHttpClient;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
+import software.amazon.awssdk.services.dynamodb.DynamoDbClientBuilder;
+import software.amazon.awssdk.services.dynamodb.model.AttributeDefinition;
+import software.amazon.awssdk.services.dynamodb.model.BillingMode;
+import software.amazon.awssdk.services.dynamodb.model.CreateTableRequest;
+import software.amazon.awssdk.services.dynamodb.model.DescribeTableRequest;
+import software.amazon.awssdk.services.dynamodb.model.DescribeTableResponse;
+import software.amazon.awssdk.services.dynamodb.model.KeySchemaElement;
+import software.amazon.awssdk.services.dynamodb.model.KeyType;
+import software.amazon.awssdk.services.dynamodb.model.ResourceNotFoundException;
+import software.amazon.awssdk.services.dynamodb.model.ScalarAttributeType;
+import software.amazon.awssdk.services.dynamodb.model.TableDescription;
+
+public class DynamoDatabaseClient implements DatabaseConnectionProvider<DynamoClientConfig> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DynamoDatabaseClient.class);
+
+  DynamoDbClient client;
+  private boolean externallyProvidedClient;
+
+  private DynamoClientConfig config;
+
+  @Override
+  public void configure(DynamoClientConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void initialize() {
+    if (this.client != null) {
+      throw new IllegalStateException("Already initialized.");
+    }
+
+    if (config instanceof DefaultDynamoClientConfig) {
+      DefaultDynamoClientConfig cfg = (DefaultDynamoClientConfig) config;
+      DynamoDbClientBuilder clientBuilder =
+          DynamoDbClient.builder()
+              .httpClient(UrlConnectionHttpClient.create())
+              .region(Region.of(cfg.getRegion()));
+
+      if (cfg.getCredentialsProvider() != null) {
+        clientBuilder = clientBuilder.credentialsProvider(cfg.getCredentialsProvider());
+      }
+      if (cfg.getEndpointURI() != null) {
+        clientBuilder = clientBuilder.endpointOverride(URI.create(cfg.getEndpointURI()));
+      }
+
+      this.externallyProvidedClient = false;
+      client = clientBuilder.build();
+    } else if (config instanceof ProvidedDynamoClientConfig) {
+      ProvidedDynamoClientConfig cfg = (ProvidedDynamoClientConfig) config;
+      this.externallyProvidedClient = true;
+      this.client = cfg.getDynamoDbClient();
+    } else {
+      throw new IllegalArgumentException(
+          "Must provide a Dynamo-client-configuration of type DefaultDynamoClientConfig or ProvidedDynamoClientConfig.");
```

Comment:
Could we restructure our setup code to avoid this `if` chain? :thinking: I'll also see about that in my Mongo PR.